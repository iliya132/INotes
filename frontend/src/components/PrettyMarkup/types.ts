export interface IPrettyMarkupProps {
    renderedValue: string;
    className?: string;
}
