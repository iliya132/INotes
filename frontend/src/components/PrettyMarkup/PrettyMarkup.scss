.workfield-content-display-field {
    margin: 5px 5px 0;
    padding: 15px;
    border: none;
    max-width: 100%;
    border-radius: 5px;
    overflow-y: auto;
    overflow-x: hidden;

    & * {
        word-wrap: break-word;
    }
}
