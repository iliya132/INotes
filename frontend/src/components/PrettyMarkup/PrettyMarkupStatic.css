:not(pre) > code {
    background-color: #d3d3d3;
    color: #353535;
    font-size: 14px;
    font-family: monospace;
}

table{
	border: 1px solid #eee;
	table-layout: fixed;
	width: 100%;
	margin-bottom: 20px;
    border-collapse: collapse;
}
table th {
	font-weight: bold;
	padding: 5px;
	background: #efefef;
	border: 1px solid #dddddd;
}
table td{
	padding: 5px 10px;
	border: 1px solid #eee;
	text-align: left;
}
table tbody tr:nth-child(odd){
	background: #fff;
}
table tbody tr:nth-child(even){
	background: #F7F7F7;
}

blockquote {
	padding: 10px 20px;
	margin: 0 0 20px;
	font-size: 17.5px;
	border-left: 5px solid #eee;
}

blockquote > p {
	margin: 0;
	margin-bottom: 1em;
}