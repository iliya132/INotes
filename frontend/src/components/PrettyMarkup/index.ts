import { PrettyMarkup } from "./PrettyMarkup";

export default PrettyMarkup;
