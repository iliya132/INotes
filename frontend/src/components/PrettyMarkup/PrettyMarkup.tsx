import React from 'react';
import { IPrettyMarkupProps } from './types';
import styles from './PrettyMarkup.scss';
import classNames from 'classnames';
import './PrettyMarkupStatic.css';
import sanitizeHtml from 'sanitize-html';

export function PrettyMarkup(props: IPrettyMarkupProps) {
    const { renderedValue, className } = props;
    const toRender = sanitize(renderedValue);
    return (
        <div
            data-testid="pretty-view"
            className={classNames(styles['workfield-content-display-field'], className)}
            dangerouslySetInnerHTML={{ __html: toRender }}></div>
    );
}

const defaults = sanitizeHtml.defaults
const config = {
    allowedTags: [...defaults.allowedTags, "img"],
    allowedAttributes: {...defaults.allowedAttributes, "*": ["class"]}
}

function sanitize(toSanitize: string): string {
    return sanitizeHtml(toSanitize, config)
}
