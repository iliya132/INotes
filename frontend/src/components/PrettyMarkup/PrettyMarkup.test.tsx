/* eslint-disable no-undef */
import React from 'react';
import { createRoot, Root } from 'react-dom/client';
import { act } from 'react-dom/test-utils';
import { PrettyMarkup } from './PrettyMarkup';

//#region common setup
let root: Root | null = null;
let container: Element | null = null;
beforeEach(() => {
    act(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
        root = createRoot(container);
    });
});

afterEach(() => {
    act(() => {
        root!.unmount();
        container!.remove();
        container = null;
    });
});
//#endregion

describe("pretty markup tests", () => {
    it("can render", () => {
        act(() => {
            root!.render(<PrettyMarkup renderedValue='<div>test-case</div>'/>)
        })
    })

    it("renders value", () => {
        act(() => {
            root!.render(<PrettyMarkup renderedValue='<div id="test-id">test-case</div>'/>)
        })

        const renderedDiv = container!.querySelector("div")

        expect(renderedDiv).not.toBeNull()
        expect(renderedDiv?.textContent).toEqual("test-case")
    })

    it("should match snapshot", () => {
        act(() => {
            root!.render(<PrettyMarkup renderedValue='<div id="test-id">test-case</div>'/>)
        })

        const markup = container!.firstChild;

        expect(markup).toMatchSnapshot("pretty markup snapshot")
    })

    it("should be sanitized", () => {
        act(() => {
            root!.render(<PrettyMarkup renderedValue='<a onClick="alert(\"shoulnt fire\")>test</a>'/>)
        })
        const link = container!.querySelector("a");

        expect(link?.outerHTML).not.toContain('alert')
        expect(link?.outerHTML).not.toContain('onClick')

        act(() => {
            root!.render(<PrettyMarkup renderedValue='<script>console.log("atached server")</script>'/>)
        })

        expect(container!.innerHTML).not.toContain('script')
    })

    it("should keep classNames", () => {
        act(() => {
            root!.render(<PrettyMarkup renderedValue='<div class="test-class">press me</div>'/>)
        })

        const div = container!.querySelector('.test-class')
        expect(div).not.toBeNull()
    })
})