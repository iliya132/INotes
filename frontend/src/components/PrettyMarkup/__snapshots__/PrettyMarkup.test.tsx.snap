// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pretty markup tests should match snapshot: pretty markup snapshot 1`] = `
<div
  class="workfield-content-display-field"
  data-testid="pretty-view"
>
  <div>
    test-case
  </div>
</div>
`;
